packages/eslint-plugin/tests/rules/prefer-promise-reject-errors.test.ts

```diff
@@ -0,0 +1,227 @@
+      new Promise(function (resolve, reject) {
+        reject(foo().err);
+      });
+    `,
+    'new Promise((resolve, reject) => reject((foo = new Error())));',
+    `
+      new Foo((resolve, reject) => reject(5));
+    `,
+    `
+      class Foo {
+        constructor(
+          executor: (resolve: () => void, reject: (reason?: any) => void) => void,
+        ): Promise<any> {}
+      }
+      new Foo((resolve, reject) => reject(5));
+    `,
+    `
+      new Promise((resolve, reject) => {
+        return function (reject) {
+          reject(5);
+        };
+      });
+    `,
+    'new Promise((resolve, reject) => resolve(5, reject));',
+    `
+      class C {
+        #error: Error;
+        foo() {
+          Promise.reject(this.#error);
+        }
+      }
+    `,
+    `
+      const foo = Promise;
+      new foo((resolve, reject) => reject(new Error()));
+    `,
+  ],
+  invalid: [
+    'Promise.reject(5)',
+    "Promise.reject('foo')",
+    'Promise.reject(`foo`)',
+    "Promise.reject('foo', somethingElse)",
+    'Promise.reject(false)',
+    'Promise.reject(void `foo`)',
+    'Promise.reject()',
+    'Promise.reject(undefined)',
+    {
+      code: 'Promise.reject(undefined)',
+      options: [{ allowEmptyReject: true }] as const,
+    },
+    'Promise.reject(null)',
+    'Promise.reject({ foo: 1 })',
+    'Promise.reject([1, 2, 3])',
+    `
+      declare const foo: Error | undefined
+      Promise.reject(foo)
+    `,
+    `
+      declare const foo: () => Promise<string>
+      Promise.reject(await foo())
+    `,
+    `
+      declare const foo: boolean
+      Promise.reject(foo && new Error())
+    `,
+    `
+      const foo = Promise
+      foo.reject()
+    `,
+
+    'Promise.reject?.(5)',
+    'Promise?.reject(5)',
+    'Promise?.reject?.(5)',
+    '(Promise?.reject)(5)',
+    '(Promise?.reject)?.(5)',
+    `Promise['reject'](5)`,
+
+    // Assignments with mathematical operators will either evaluate to a primitive value or throw a TypeError
+    'Promise.reject(foo += new Error())',
+    'Promise.reject(foo -= new Error())',
+    'Promise.reject(foo **= new Error())',
+    'Promise.reject(foo <<= new Error())',
+    'Promise.reject(foo |= new Error())',
+    'Promise.reject(foo &= new Error())',
+
+    'new Promise(function(resolve, reject) { reject() })',
+    'new Promise(function(resolve, reject) { reject(5) })',
+    'new Promise((resolve, reject) => { reject() })',
+    'new Promise((resolve, reject) => reject(5))',
+    `
+      new Promise((resolve, reject) => {
+        fs.readFile('foo.txt', (err, file) => {
+          if (err) reject('File not found')
+          else resolve(file)
+        })
+      })
+    `,
+    'new Promise((yes, no) => no(5))',
+    'new Promise(({foo, bar, baz}, reject) => reject(5))',
+    'new Promise(function(reject, reject) { reject(5) })',
+    'new Promise(function(foo, arguments) { arguments(5) })',
+    'new Promise((foo, arguments) => arguments(5))',
```

Comment:
hmm, WDYT about this one (in both valid and invalid cases)?
```TS
'new Promise((...args) => args[1](5))',
```

Reply:
I thought about it, but I think we can do it in one of the following prs to make it easier to review the current one

Also, if we add support for this case, we would probably need to support a plenty of similar cases such as:

```tsx
new Promise((args) => { const [,reject] = args; reject(); })
```

```tsx
let reject
new Promise((resolve, _reject) => reject = _reject)
reject()
```

```tsx
new Promise(resolve, _reject) => {
  reject = _reject
  reject()
})
```
and so on...

Reply:
Yeah I don't think we need to check these kinds of `...args` cases 🙂 that's rather more complex than what most rules try to handle.